lib/vector-core/src/sink.rs

```diff
@@ -60,6 +76,96 @@ impl fmt::Debug for VectorSink {
 // === StreamSink ===
 
 #[async_trait]
-pub trait StreamSink {
-    async fn run(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()>;
+pub trait StreamSink<T> {
```

Comment:
My assumption for why we had `BoxStream` originally was because it might be difficult to add a generic param to this trait. That's clearly not true. This might be out of scope for this PR, but we should consider making the entire stream generic to remove the Boxing
```rust
pub trait StreamSink<T: Stream<Item=T>> {
    async fn run(self: Box<Self>, input: T) -> Result<(), ()>;
}
```

Reply:
That's a good thought. I'll look into converting that.

Reply:
Actually, I'm not sure if that's possible, or at least not a quick change. There are a number of different stream types that are passed to `VectorSink::run`, so they need to be boxed to keep the `StreamSink` trait object safe.